application/transaction.go

```diff
@@ -18,21 +26,33 @@ func Transactional(db DB, todo func(f Application) error) error {
 		return errors.WithStack(err)
 	}
 
-	defer func() {
-		switch err {
-		case nil:
-			log.Debug(nil, map[string]interface{}{}, "Commit the transaction!")
-			err = tx.Commit()
-		default:
-			log.Debug(nil, map[string]interface{}{}, "Rolling back the transaction...")
-			_ = tx.Rollback()
-			log.Error(nil, map[string]interface{}{
-				"err": err,
-			}, "database transaction failed!")
-			err = errors.WithStack(err)
+	return func() error {
+		errorChan := make(chan error, 1)
+		txTimeout := time.After(databaseTransactionTimeout)
+
+		go func(tx Transaction) {
+			errorChan <- todo(tx)
+		}(tx)
+
+		select {
+		case err := <-errorChan:
+			if err != nil {
+				log.Debug(nil, nil, "Rolling back the transaction...")
+				tx.Rollback()
+				log.Error(nil, map[string]interface{}{
+					"err": err,
+				}, "database transaction failed!")
+				return errors.WithStack(err)
+			}
+
+			tx.Commit()
+			log.Debug(nil, nil, "Commit the transaction!")
+			return nil
+		case <-txTimeout:
+			log.Debug(nil, nil, "Rolling back the transaction...")
+			tx.Rollback()
+			log.Error(nil, nil, "database transaction timeout!")
+			return errors.New("database transaction timeout!")
```

Comment:
> The only thing I worry about is that the go routine from above keeps on running even after having rolled back the transaction.
or should we add a `context.Context` parameter in the `todo(tx)` function to give the underlying code a chance to cancel its work ? Also, I assume that at the repository level, if the `tx` is rolled-back while a SQL query is running, their will be an error and the go routine will stop, right ?